### PR TITLE
Add Auth Code Client, Persisted Grants. Support Head Requests

### PR DESCRIPTION
**Pull request by bderveni**
- Add Authorization Code Clients
- Add Persisted Grants
- Re-write some of the other docs
- Support HEAD requests

### CODE REVIEW

**Comment by bderveni** [comment]
@osimarkrogers

**Comment by bderveni** [comment]
@mckenzielee would appreciate your feedback on this PR.

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+uid: identityAuthorizationCodeClient
+---
+
+# Authorization Code Client
+
+Authorization Code clients are used in Javascript/Browser (SPA) based applications or native
+            mobile applications with the presence of a User.
+            These clients are issued an ID. You can read more about these clients
+            [here](https://github.com/osisoft/OSI-Samples/tree/master/ocs_samples/basic_samples/Authentication#authorization-code-flow-with-pkce).
+            These clients are not issued secrets or refresh tokens.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Replace "These" with "Authorization Code"

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+
+## Properties
+
+For HTTP requests and responses, the AuthorizationCodeClient object has the following properties and JSON-serialized body: 
+
+Property | Type | Descriptions
+ --- | --- | --- | ---
+AllowedCorsOrigins | string[] | If specified, will be used by the default CORS policy service implementations to build a CORS policy for JavaScript clients.
+RedirectUris | string[] | Specifies the allowed URIs to return tokens or authorization codes to. Wildcards are ignored. URIs must match exactly to what you are redirecting to after login. If URIs do not match the authentication process will fail with a bad_client error. Maximum 10 per client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Add comma after "not match"

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
+
+`Authorization: Bearer <token>`
+
+Requests made without an access token, or an invalid/expired one will fail with a 401 Unauthorized response.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change "invalid/expired one" to "invalid/expired token"

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+Requests made with an access token, which does not have the correct permissions (see security subsection on every endpoint) will fail with a 403 Forbidden.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Delete comma

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+[Required]
+AuthorizationCodeClient authorizationCodeClient
+```
+
+New AuthorizationCodeClient object.
+
+Property | Type | Required | Description 
+ --- | --- | --- | ---
+AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.
+RedirectUris | string[] | No | Specifies the allowed URIs to return tokens or authorization codes to.            Wildcards are ignored. URIs must match exactly to what you are redirecting            to after login. If URIs do not match the authentication process will fail            with a bad_client error.            Maximum 10 per client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change to "Specifies the allowed URIs to which return tokens or authorization codes can be returned."

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 6]
```diff
@@ -0,0 +1,784 @@
...
+[Required]
+AuthorizationCodeClient authorizationCodeClient
+```
+
+Updated Authorization Code Client values. Properties that are not set or are null will not be changed.
+
+Property | Type | Required | Description 
+ --- | --- | --- | ---
+AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.
+RedirectUris | string[] | No | Specifies the allowed URIs to return tokens or authorization codes to.            Wildcards are ignored. URIs must match exactly to what you are redirecting            to after login. If URIs do not match the authentication process will fail            with a bad_client error.            Maximum 10 per client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change to "Specifies the allowed URIs to which return tokens or authorization codes can be returned."

Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change "must match exactly to what you are redirecting to..." to "exactly to what you are redirecting..."

File: Documentation/Identity/Identity_AuthorizationCodeClient.md (hunk 5 above)
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change "must match exactly to what you are redirecting to..." to "exactly to what you are redirecting..."

File: Documentation/Identity/Identity_AuthorizationCodeClient.md [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+PostLogoutRedirectUris | string[] | No | Specifies allowed URIs to redirect to after logout. Wildcards are ignored.            URIs must match exactly to what you are redirecting to after logout.            Maximum 10 for client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AuthorizationCodeClient.md]: Change "must match exactly to what you are redirecting to..." to "exactly to what you are redirecting after logout."

File: Documentation/Identity/Identity_AzureActiveDirectoryTenant.md [hunk 8]
```diff
@@ -328,29 +365,29 @@ Send consent for an Azure Active Directory Tenant
...
-Consent information used to send
+ConsentInformation object.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-AzureActiveDirectoryConsentEmail | string | Yes | Preferred Azure Active Directory consent email for user.
-AzureActiveDirectoryConsentGivenName | string | Yes | Preferred Azure Active Directory consent name for user.
-AzureActiveDirectoryConsentSurname | string | Yes | Preferred Azure Active Directory consent surname for user.
-AzureActiveDirectoryTenant | string | Yes | Domain Name of Azure Active Directory (e.g. mydomain.onmicrosoft.com)
+AzureActiveDirectoryConsentEmail | string | Yes | Email to send the consent email to.            Only AAD Admins have permission to consent to OCS            being allowed to interact with the tenant. The email            does not have to be sent to an Admin.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AzureActiveDirectoryTenant.md]: Change "Email to send the consent email to." to "Address to email consent."

File: Documentation/Identity/Identity_ClientCredentialClient.md [hunk 9]
```diff
@@ -38,9 +48,44 @@ Tags | string[] | For OSIsoft internal use only
...
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
+
+`Authorization: Bearer <token>`
+
+Requests made without an access token, or an invalid/expired one will fail with a 401 Unauthorized response.
+Requests made with an access token, which does not have the correct permissions (see security subsection on every endpoint) will fail with a 403 Forbidden.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ClientCredentialClient.md]: Delete comma after "access token"

File: Documentation/Identity/Identity_ClientCredentialClient.md [hunk 10]
```diff
@@ -53,39 +98,41 @@ Create a Client Credential flow Client
...
-New ClientCredentialClientCreate object
+ClientCredentialClientCreate object.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-SecretDescription | string | No | Description for the initial secret for the client.
-SecretExpirationDate | DateTime | No | Expiration date for the initial secret for the client.
-RoleIds | Guid[] | No | List of Roles to be assigned to this client.
-Id | string | No | Client ID for this Client.Must be the same as the one in the route
+SecretDescription | string | No | Description for the initial secret for the client. Ensure that this is descriptive enough            as it will be the only way to distinguish between multiple secrets and their usage for a            client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ClientCredentialClient.md]: Need comma after "descriptive enough"

File: Documentation/Identity/Identity_HybridClient.md [hunk 11]
```diff
@@ -2,26 +2,35 @@
...
-Property | Type | Description
- --- | --- | ---
-AllowOfflineAccess | bool | Specifies whether this client can request refresh tokens, by providing the *offline_access* scopes.
+Property | Type | Descriptions
+ --- | --- | --- | ---
+AllowOfflineAccess | bool | Specifies whether this client can request refresh tokens, by providing the *offline_access* scope.
 AllowAccessTokensViaBrowser | bool | Specifies whether this HybridClient is allowed to receive access tokens via the browser. This is useful to harden flows that allow multiple response types (e.g. by disallowing a hybrid flow client that is supposed to use code *id_token* to add the *token* response type, thus leaking the token to the browser).
-RedirectUris | string[] | Specifies the allowed URIs to return tokens or authorization codes to.
-PostLogoutRedirectUris | string[] | Specifies allowed URIs to redirect to after logout.
+RedirectUris | string[] | Specifies the allowed URIs to return tokens or authorization codes to. Wildcards are ignored. URIs must match exactly to what you are redirecting to after login. If URIs do not match the authentication process will fail with a bad_client error. Maximum 10 per client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_HybridClient.md]: delete "to" after "redirecting"

File: Documentation/Identity/Identity_HybridClient.md [hunk 12]
```diff
@@ -51,9 +61,44 @@ Tags | string[] | For OSIsoft internal use only
...
 
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_HybridClient.md]: Add space between these two sentences.  Add "The" to "Authorization header"

File: Documentation/Identity/Identity_AzureActiveDirectoryTenant.md [hunk 13]
```diff
@@ -26,9 +29,40 @@ ConsentState | ConsentState | Consent State of Azure Active Directory Tenant. Ca
...
-## `Add Azure Active Directory Tenant`
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_AzureActiveDirectoryTenant.md]: Add space between these two sentences.  Add "The" to "Authorization header"

File: Documentation/Identity/Identity_ClientCredentialClient.md [hunk 14]
```diff
@@ -38,9 +48,44 @@ Tags | string[] | For OSIsoft internal use only
...
 
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ClientCredentialClient.md]: Add space between these two sentences. Add "The" to "Authorization header"

File: Documentation/Identity/Identity_HybridClient.md [hunk 15]
```diff
@@ -66,37 +111,38 @@ Create a Hybrid flow Client
...
 
-New HybridClientCreate object
+HybridClientCreate object.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
 SecretDescription | string | No | Description for the initial secret for the client.
-SecretExpirationDate | DateTime | No | Expiration date for the initial secret for the client.
-AllowOfflineAccess | bool | No | Specifies whether this client can request refresh tokens, by providing the *offline_access* scopes.
+SecretExpirationDate | DateTime | No | Expiration date for the initial secret for the client. If set to null the secret will            never expire. We advise against such practice.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_HybridClient.md]: Extra spaces from line 127 to 138 - - will these show up in Doc FX?

File: Documentation/Identity/Identity_IdentityProvider.md [hunk 16]
```diff
@@ -30,6 +30,37 @@ UserIdClaimType | string | Type of claim
...
 
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_IdentityProvider.md]: Add space between these two sentences. Add "The" to "Authorization header"

File: Documentation/Identity/Identity_IdentityProvider.md [hunk 17]
```diff
@@ -266,24 +529,24 @@ Add an existing identity provider to a tenant
...
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
 IdentityProviderId | Guid | Yes | Identity Provider Id to Add
 AzureActiveDirectorySendConsent | bool | No | Send consent email for Azure Active Directory.
-AzureActiveDirectoryConsentEmail | string | No | Preferred Azure Active Directory consent email for user. Must be provided if AzureActiveDirectorySendConsent is set to true.
-AzureActiveDirectoryConsentGivenName | string | No | Preferred Azure Active Directory consent name for user. Must be provided if AzureActiveDirectorySendConsent is set to true.
-AzureActiveDirectoryConsentSurname | string | No | Preferred Azure Active Directory consent surname for user. Must be provided if AzureActiveDirectorySendConsent is set to true.
-AzureActiveDirectoryTenant | string | No | Domain Name of Azure Active Directory (e.g. mydomain.onmicrosoft.com). Must be provided if you add AAd as an Identity Provider.
+AzureActiveDirectoryConsentEmail | string | Yes | Email to send the consent email to.            Only AAD Admins have permission to consent to OCS            being allowed to interact with the tenant. The email            does not have to be sent to an Admin.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_IdentityProvider.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_HybridClient.md (hunk 15 above)
Comment (bderveni) [reply Documentation/Identity/Identity_HybridClient.md]: They will not

File: Documentation/Identity/Identity_IdentityProvider.md [hunk 18]
```diff
@@ -325,32 +588,34 @@ Created
...
 
-Internal server error
+Internal server error.
 ***
 
-## `Remove Identity Provider From Tenant`
+## `Remove IDP From Tenant`
 
-Remove an identity provider from a tenant
+Remove an Identity Provider from a Tenant. Users provisioned
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_IdentityProvider.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_IdentityProvider.md [hunk 19]
```diff
@@ -382,22 +647,136 @@ Allowed for these roles:
...
+IdentityProvider or Tenant not found.
+
+#### 500
+
+Internal server error.
+***
+
+## `Validate IDP Exists`
+
+Validate that a Identity Provider exists in the Tenant.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_IdentityProvider.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_ClientCredentialClient.md (hunk 14 above)
Comment (bderveni) [reply Documentation/Identity/Identity_ClientCredentialClient.md]: Updated on another commit.

File: Documentation/Identity/Identity_IdentityProvider.md (hunk 17 above)
Comment (bderveni) [reply Documentation/Identity/Identity_IdentityProvider.md]: Nope.

File: Documentation/Identity/Identity_ImplicitClient.md [hunk 20]
```diff
@@ -52,9 +59,41 @@ Tags | string[] | For OSIsoft internal use only
...
-## `Create Implicit Client`
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ImplicitClient.md]: Add space between these two sentences. Add "The" to "Authorization header"

File: Documentation/Identity/Identity_IdentityProvider.md (hunk 19 above)
Comment (bderveni) [reply Documentation/Identity/Identity_IdentityProvider.md]: No.

File: Documentation/Identity/Identity_ImplicitClient.md [hunk 21]
```diff
@@ -67,27 +106,28 @@ Create an Implicit flow Clients
...
 
-New ImplicitClient object
+New ImplicitClient object.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.
-RedirectUris | string[] | No | Specifies the allowed URIs to return tokens or authorization codes to.
-PostLogoutRedirectUris | string[] | No | Specifies allowed URIs to redirect to after logout.
+AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.            Maximum 10 for client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ImplicitClient.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_ImplicitClient.md [hunk 22]
```diff
@@ -199,34 +242,35 @@ Update an Implicit Client
...
 
-Updated Implicit Client values
+Updated Implicit Client values. Properties that are not set or are null will not be changed.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.
-RedirectUris | string[] | No | Specifies the allowed URIs to return tokens or authorization codes to.
-PostLogoutRedirectUris | string[] | No | Specifies allowed URIs to redirect to after logout.
+AllowedCorsOrigins | string[] | No | If specified, will be used by the default CORS policy service implementations to build a            CORS policy for JavaScript clients.            Maximum 10 for client.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ImplicitClient.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_ImplicitClient.md [hunk 23]
```diff
@@ -524,24 +584,26 @@ Success
...
 
-Internal server error
+Internal server error.
 ***
 
-## `Delete Implicit Client`
+## `Delete an Implicit Client`
 
-Delete an Implicit Client
+Delete an Implicit Client. It can take up to one hour
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_ImplicitClient.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_Invitation.md [hunk 24]
```diff
@@ -4,14 +4,25 @@ uid: identityInvitation
 
 # Invitation
 
-APIs for creating, getting, updating, and deleting an Invitation
+Invitations are issued after the creation of a User object in OCS,
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Extra spaces?

File: Documentation/Identity/Identity_Invitation.md [hunk 25]
```diff
@@ -38,9 +49,40 @@ IdentityProviderId | Guid | ID of the Identity Provider that must be used to acc
...
 
+## Base URL
+
+All URLs referenced in this section have the following base:
+
+`https://dat-b.osisoft.com/`
+
+## Authentication
+
+All endpoints referenced in this documentation require authenticated access.Authorization header must be set to the access token you retrieve after a successful authentication request.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Add space between these two sentences. Add "The" to "Authorization header"

File: Documentation/Identity/Identity_Invitation.md [hunk 26]
```diff
@@ -238,35 +281,35 @@ Update an invitation
...
-New InvitationCreateOrUpdate object
+New InvitationCreateOrUpdate object. Properties that are not set or are null will not be changed.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-ExpiresDateTime | DateTime | No | Invitation expiration date. Must be in the future.
-State | InvitationStates | No | Set the state of invitation.
-SendInvitation | bool | No | Send the invitation. Default is true.
-IdentityProviderId | Guid | No | Identity Provider to use for accepting this invitation. Null implies invitation can be accepted using any            configured Identity Provider.
+ExpiresDateTime | DateTime | No | Invitation expiration date. Must be in the future.            Maximum allowed is two month in the future.            Defaults to 21 days on creation.            It should be in [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) and either            include a *Z* at the end to represent UTC timezone            or include the offset in hours. If neither is present            time will be treated in the local time zone of the server.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_Invitation.md [hunk 27]
```diff
@@ -357,28 +401,151 @@ Allowed for these roles:
...
+Invitation or Tenant not found.
+
+#### 500
+
+Internal server error.
+***
+
+## `Get header for Invitations`
+
+Return total number of non-expired Invitations in a Tenant. Optionally include expired invitations.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Extra spaces?

File: Documentation/Identity/Identity_Invitation.md [hunk 28]
```diff
@@ -461,35 +629,35 @@ Create an invitation for a user. Should use when no other invitation exists for
...
-InvitationCreateOrUpdate object
+InvitationCreateOrUpdate object.
 
 Property | Type | Required | Description 
  --- | --- | --- | ---
-ExpiresDateTime | DateTime | No | Invitation expiration date. Must be within 2 months in the future.
-State | InvitationStates | No | Set the state of invitation.
-SendInvitation | bool | No | Send the invitation. Default is true.
-IdentityProviderId | Guid | No | Identity Provider to use for accepting this invitation. Null implies invitation can be accepted using any            configured Identity Provider.
+ExpiresDateTime | DateTime | No | Invitation expiration date. Must be in the future.            Maximum allowed is two month in the future.            Defaults to 21 days on creation.            It should be in [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) and either            include a *Z* at the end to represent UTC timezone            or include the offset in hours. If neither is present            time will be treated in the local time zone of the server.
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Extra spaces will show up in DocFX?

File: Documentation/Identity/Identity_Invitation.md [hunk 29]
```diff
@@ -565,35 +733,35 @@ Create or update an invitation for a user
...
```
Comment (osimarkrogers) [inline Documentation/Identity/Identity_Invitation.md]: Extra spaces?

**Pull request merged**